File: src/main/java/io/trydent/olimpo/Olimpo.java
package io.trydent.olimpo;

import static io.trydent.olimpo.action.Action.commandAction;
import static io.trydent.olimpo.apollo.ApolloCommand.addReading;
import static io.trydent.olimpo.db.DbmsParams.envPostgresql;
import static io.trydent.olimpo.db.SqlClient.dbmsClient;
import static io.trydent.olimpo.http.HttpExchange.actionSwitch;
import static io.trydent.olimpo.http.HttpExchange.staticContent;
import static io.trydent.olimpo.http.HttpRequest.routeSwitch;
import static io.trydent.olimpo.http.HttpRoute.action;
import static io.trydent.olimpo.http.HttpRoute.webroot;
import static io.trydent.olimpo.http.HttpServer.httpServer;
import static io.trydent.olimpo.io.Port.envPort;
import static io.trydent.olimpo.sink.CommandSink.commandBus;
import static io.trydent.olimpo.sys.Id.uuid;
import static io.trydent.olimpo.vertx.Deployment.verticles;

public final class Olimpo {
  private Olimpo() {
  }

  public static void main(String... args) {
    verticles(
      vertx -> httpServer(
        vertx,
        routeSwitch(
          vertx,
          webroot("/*",
            staticContent(
              "webroot"
            )
          ),
          action("/apollo/:action",
            actionSwitch(
              commandAction(vertx.eventBus())
            )
          )
        )
      ).listen(envPort("PORT", 8095)),
      vertx -> commandBus(vertx.eventBus(), uuid())
        .let("add-reading",
          addReading(
            dbmsClient(
              vertx,
              envPostgresql("DATABASE_URL")
            )
          )
        )
    ).deploy();
  }
}

File: src/main/java/io/trydent/olimpo/Olimpo.kt
package io.trydent.olimpo

import io.trydent.olimpo.action.Action.Companion.busCommand
import io.trydent.olimpo.apollo.ApolloCommand.Companion.addReading
import io.trydent.olimpo.bus.Bus.Companion.commandBus
import io.trydent.olimpo.bus.Process.Companion.commandProcess
import io.trydent.olimpo.http.HttpExchange.Companion.actionSwitch
import io.trydent.olimpo.http.HttpExchange.Companion.staticContent
import io.trydent.olimpo.http.HttpRequest.Companion.routeSwitch
import io.trydent.olimpo.http.HttpRoute.Companion.action
import io.trydent.olimpo.http.HttpRoute.Companion.webroot
import io.trydent.olimpo.http.HttpServer.Companion.httpServer
import io.trydent.olimpo.io.Port.Companion.envPort
import io.trydent.olimpo.sys.Id.Companion.uuid
import io.trydent.olimpo.vertx.deploy
import io.vertx.core.Vertx.vertx

fun main() = vertx().deploy({ vertx ->
  httpServer(
    vertx,
    routeSwitch(
      vertx,
      webroot("/*",
        staticContent(
          "webroot"
        )
      ),
      action("/apollo/:action",
        actionSwitch(
          busCommand(vertx.eventBus())
        )
      )
    )
  )(
    envPort(name = "PORT", default = 8095)
  )
}, { vertx ->
  commandBus(
    commandProcess(uuid(), vertx.eventBus()),
    addReading()
  )()
})
